python/lsst/analysis/drp/calcFunctors.py

```diff
@@ -143,3 +143,180 @@ def columns(self):
     def __call__(self, df):
         size = np.power(df[self.colXx]*df[self.colYy] - df[self.colXy]**2, 0.25)
         return size
+
+
+class ColorDiff(MultiColumnAction):
+    """Calculate the difference between two colors;
+    each color is derived from two flux columns.
+
+    The color difference is computed as (color1 - color2) with:
+
+    color1 = color1_mag1 - color1_mag2
+    color2 = color2_mag1 - color2_mag2
+
+    where color1_mag1 is the magnitude associated with color1_flux1, etc.
+
+    Parameters
+    ----------
+    df : `pandas.core.frame.DataFrame`
+        The catalog to calculate the color difference from.
+
+    Returns
+    -------
+    The color difference in millimags.
+
+    Notes
+    -----
+    The flux columns need to be in units that can be converted
+    to janskies. This action doesn't have any calibration
+    information and assumes that the fluxes are already
+    calibrated.
+    """
+    color1_flux1 = Field(doc="Column for flux1 to determine color1",
+                         dtype=str)
+    color1_flux1_units = Field(doc="Units for color1_flux1",
+                               dtype=str,
+                               default="nanojansky")
+    color1_flux2 = Field(doc="Column for flux2 to determine color1",
+                         dtype=str)
+    color1_flux2_units = Field(doc="Units for color1_flux2",
+                               dtype=str,
+                               default="nanojansky")
+    color2_flux1 = Field(doc="Column for flux1 to determine color2",
+                         dtype=str)
+    color2_flux1_units = Field(doc="Units for color2_flux1",
+                               dtype=str,
+                               default="nanojansky")
+    color2_flux2 = Field(doc="Column for flux2 to determine color2",
+                         dtype=str)
+    color2_flux2_units = Field(doc="Units for color2_flux2",
+                               dtype=str,
+                               default="nanojansky")
+    return_millimags = Field(doc="Use millimags or not?",
+                             dtype=bool,
+                             default=True)
+
+    @property
+    def columns(self):
+        return (self.color1_flux1,
+                self.color1_flux2,
+                self.color2_flux1,
+                self.color2_flux2)
+
+    def __call__(self, df):
+        color1_flux1 = df[self.color1_flux1].values*u.Unit(self.color1_flux1_units)
+        color1_mag1 = color1_flux1.to(u.ABmag).value
+
+        color1_flux2 = df[self.color1_flux2].values*u.Unit(self.color1_flux2_units)
+        color1_mag2 = color1_flux2.to(u.ABmag).value
+
+        color2_flux1 = df[self.color2_flux1].values*u.Unit(self.color2_flux1_units)
+        color2_mag1 = color2_flux1.to(u.ABmag).value
+
+        color2_flux2 = df[self.color2_flux2].values*u.Unit(self.color2_flux2_units)
+        color2_mag2 = color2_flux2.to(u.ABmag).value
+
+        color1 = color1_mag1 - color1_mag2
```

Comment:
This is fine for now, but we should consider this as a technical debt and refactor it later to have a functor/action to calculate colors and use the `DifferenceColumns` action to compute color difference. I also didn't look at it in too much detail, but wonder if these can be achieved through the data frame actions [here](https://github.com/lsst/pipe_tasks/blob/13b7538eaf671293994b26784b4df6d834d07738/python/lsst/pipe/tasks/dataFrameActions/_actions.py#L60).

Reply:
I'm curious why using a composite action might be considered better here than a simple subtraction - I am very new to actually using these, so there may be something important I am not aware of, but I worry that this is a case where maximizing use of something we wrote is automatically assumed to be good.

I think the actual criteria for using ConfigurableAction composition instead of writing a non-composite action should be something like this:

- Would there actually be any use cases for swapping out the nested action?
- If so, is it also important for those other use cases to share the parent action from a correctness standpoint (i.e. if we had two non-composite actions instead, and they got out of sync, is that bug)?

For cases where it's just that the nested action is complex enough that we want to encourage it to be reused, I think I'd rather see that complexity moved to other normal classes or functions that could be reused instead.

I'm thinking we might want to raise this at a team meeting; maybe try to brainstorm some conventions and best practices and get them written down.

Reply:
My thought was that computing a color given a pair of bands would be a common thing to do, and hence we should have an action that would take in `"g"` and `"r"` say and give colors. I am okay with either subtracting the color after that to compute `ColorDiff` or use `DifferenceColumn` action.

Reply:
I think that this is definitely touching on something philosophical that we should discuss in team meeting.  I am not a fan of the nested actions, actually.  I much prefer programming in python than trying to turn yaml/pexConfig into a programming language.
The way that the plotting tools can be reused and configured to do different actions for each column is great.  But I don't see the problem in each column type being a particular block of python code.

Reply:
Yes, topic for a team meeting for sure!